### DISCUSSION TITLE
Simulating robot with closed-clain mechanism

### DISCUSSION THREAD

**Discussion by nisal-perera** [google-deepmind/mujoco#2268] [Asking for Help]
### Intro

Hi!

I am a PhD student using Mujoco to simulate a bipedal robot with closed chain mechanism.

### My setup

I am using Mujoco 3.2.4 on ubuntu 20.04 with C++.

### My question

I made a mujoco model for my biped robot and created the closed-chain mechanism using equality (`<connect>`) constraint. I have attached a simplified version of the robot model. As you can see, the closed chain mechanism is a bit complicated having spatial 6-bar and 5-bar linkages. I am trying to implement a whole-body controller to control the posture of the robot. Using the WBC, I am applying torques to the pulleys that drive the mechanism and actuate the joints. However, I found it difficult to stabilize the system by applying the torques to the pulleys and it seems that it produce larger reaction force than it needed causing the instability. The WBC works with the open chain model and the command torques to the pulleys make sense as well. I was curious what are the best guidelines/settings for simulating a system like this in mujoco to verify if i have set up the simulation properly.

I am using low `solref` in `<equality>` constraint to ensure that loop closure constraints are not violated. Also, I have tried using much smaller _timesteps_ (0.0001 and 0.00001) along with `RK4` and `Euler` integration. 

### Minimal model and/or code that explain my question

[closed_chain.zip](https://github.com/user-attachments/files/18001492/closed_chain.zip)

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
Ya, stabilizing linkages is hard 🙂 

I took a quick look at your model it is indeed pretty complicated. One big issue that's obvious is that your inertias are all wrong (visualize them in simulate and you'll see what I mean).

One quite useful option is to replace bodies+connect with spatial tendons. If one of the links in your linkage is very light and doesn't need to collide with things, you can just replace it with tendon. These constraints are both cheaper and much more stable than connects. The price you pay for this is that the tendon cannot collide with things, but maybe it never should anyway.

PS why is you extent so big? It makes it impossible to zoom in and look closely at things...

**Reply by nisal-perera** [reply]
Hi Yuval,

Thanks for getting back to me. For the reduced model I provided, I just put random inertias because I just wanted to give an idea about the structure of the model and constraints. I will give using spatial tendons a try as you suggested. Sorry about the extent being big in the provided model.

**Reply by chunantian** [reply]
> Hi Yuval,
> 
> Thanks for getting back to me. For the reduced model I provided, I just put random inertias because I just wanted to give an idea about the structure of the model and constraints. I will give using spatial tendons a try as you suggested. Sorry about the extent being big in the provided model.

Hi Nisal-perera,

Do you try it successfully? I am also solving the closed chains model in Mujoco. If you were success, I hope to receive your response.
